#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [2]:
import pandas as pd

In [3]:
pip install kaggle

In [4]:
! mkdir ~/.kaggle

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
!cp /content/drive/MyDrive/kaggle.json ~/.kaggle/kaggle.json

In [7]:
! kaggle datasets download arnabchaki/data-science-salaries-2023

  0% 0.00/25.4k [00:00<?, ?B/s]
100% 25.4k/25.4k [00:00<00:00, 36.4MB/s]


In [8]:
! unzip data-science-salaries-2023.zip

Archive:  data-science-salaries-2023.zip
  inflating: ds_salaries.csv         


In [9]:
df = pd.read_csv("ds_salaries.csv")
df

work_year experience_level employment_type                 job_title  \
0          2023               SE              FT  Principal Data Scientist   
1          2023               MI              CT               ML Engineer   
2          2023               MI              CT               ML Engineer   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
...         ...              ...             ...                       ...   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

       salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0       80000             EUR          85847                 ES           100   
1       30000             USD          30000                 US           100   
2       25500             USD          25500                 US           100   
3      175000             USD         175000                 CA           100   
4      120000             USD         120000                 CA           100   
...       ...             ...            ...                ...           ...   
3750   412000             USD         412000                 US           100   
3751   151000             USD         151000                 US           100   
3752   105000             USD         105000                 US           100   
3753   100000             USD         100000                 US           100   
3754  7000000             INR          94665                 IN            50   

     company_location company_size  
0                  ES            L  
1                  US            S  
2                  US            S  
3                  CA            M  
4                  CA            M  
...               ...          ...  
3750               US            L  
3751               US            L  
3752               US            S  
3753               US            L  
3754               IN            L  

[3755 rows x 11 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [10]:
#Колонки
df.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')

In [12]:
#Первые 5 значений
df[:5]

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [13]:
#Последние 5 значений
df[-5:]

work_year experience_level employment_type                 job_title  \
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

       salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
3750   412000             USD         412000                 US           100   
3751   151000             USD         151000                 US           100   
3752   105000             USD         105000                 US           100   
3753   100000             USD         100000                 US           100   
3754  7000000             INR          94665                 IN            50   

     company_location company_size  
3750               US            L  
3751               US            L  
3752               US            S  
3753               US            L  
3754               IN            L

In [14]:
#Статистику
df.describe(include="all")

work_year experience_level employment_type      job_title  \
count   3755.000000             3755            3755           3755   
unique          NaN                4               4             93   
top             NaN               SE              FT  Data Engineer   
freq            NaN             2516            3718           1040   
mean    2022.373635              NaN             NaN            NaN   
std        0.691448              NaN             NaN            NaN   
min     2020.000000              NaN             NaN            NaN   
25%     2022.000000              NaN             NaN            NaN   
50%     2022.000000              NaN             NaN            NaN   
75%     2023.000000              NaN             NaN            NaN   
max     2023.000000              NaN             NaN            NaN   

              salary salary_currency  salary_in_usd employee_residence  \
count   3.755000e+03            3755    3755.000000               3755   
unique           NaN              20            NaN                 78   
top              NaN             USD            NaN                 US   
freq             NaN            3224            NaN               3004   
mean    1.906956e+05             NaN  137570.389880                NaN   
std     6.716765e+05             NaN   63055.625278                NaN   
min     6.000000e+03             NaN    5132.000000                NaN   
25%     1.000000e+05             NaN   95000.000000                NaN   
50%     1.380000e+05             NaN  135000.000000                NaN   
75%     1.800000e+05             NaN  175000.000000                NaN   
max     3.040000e+07             NaN  450000.000000                NaN   

        remote_ratio company_location company_size  
count    3755.000000             3755         3755  
unique           NaN               72            3  
top              NaN               US            M  
freq             NaN             3040         3153  
mean       46.271638              NaN          NaN  
std        48.589050              NaN          NaN  
min         0.000000              NaN          NaN  
25%         0.000000              NaN          NaN  
50%         0.000000              NaN          NaN  
75%       100.000000              NaN          NaN  
max       100.000000              NaN          NaN

Ответьте на следующие вопросы:

1.Какие средние значения в каждой числовой колонке вашего датасета?
2.Минимальные и максимальные значения в каждой числовой колонке вашего датасета

mean 2022.373635	NaN	NaN	NaN	1.906956e+05	NaN	137570.389880	NaN	46.271638	NaN	NaN

min 2020.000000	NaN	NaN	NaN	6.000000e+03	NaN	5132.000000	NaN	0.000000	NaN	NaN

max 2023.000000	NaN	NaN	NaN	3.040000e+07	NaN	450000.000000	NaN	100.000000	NaN	NaN

##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [16]:
df.employment_type = df.employment_type.astype("category")
df.salary	 = df.salary.astype("Float64")
df.work_year = df.work_year.astype("Int64")

#Вывести типы колонок тут
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   work_year           3755 non-null   Int64   
 1   experience_level    3755 non-null   object  
 2   employment_type     3755 non-null   category
 3   job_title           3755 non-null   object  
 4   salary              3755 non-null   Float64 
 5   salary_currency     3755 non-null   object  
 6   salary_in_usd       3755 non-null   int64   
 7   employee_residence  3755 non-null   object  
 8   remote_ratio        3755 non-null   int64   
 9   company_location    3755 non-null   object  
 10  company_size        3755 non-null   object  
dtypes: Float64(1), Int64(1), category(1), int64(2), object(6)
memory usage: 304.7+ KB


Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [17]:
#Фильтр колонки 1
df[df["salary_currency"]== "USD"]

work_year experience_level employment_type                 job_title  \
1          2023               MI              CT               ML Engineer   
2          2023               MI              CT               ML Engineer   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
5          2023               SE              FT         Applied Scientist   
...         ...              ...             ...                       ...   
3749       2021               SE              FT           Data Specialist   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   

        salary salary_currency  salary_in_usd employee_residence  \
1      30000.0             USD          30000                 US   
2      25500.0             USD          25500                 US   
3     175000.0             USD         175000                 CA   
4     120000.0             USD         120000                 CA   
5     222200.0             USD         222200                 US   
...        ...             ...            ...                ...   
3749  165000.0             USD         165000                 US   
3750  412000.0             USD         412000                 US   
3751  151000.0             USD         151000                 US   
3752  105000.0             USD         105000                 US   
3753  100000.0             USD         100000                 US   

      remote_ratio company_location company_size  
1              100               US            S  
2              100               US            S  
3              100               CA            M  
4              100               CA            M  
5                0               US            L  
...            ...              ...          ...  
3749           100               US            L  
3750           100               US            L  
3751           100               US            L  
3752           100               US            S  
3753           100               US            L  

[3224 rows x 11 columns]

In [18]:
#Фильтр колонки 2
df[df["salary"] >= 80000]

work_year experience_level employment_type                 job_title  \
0          2023               SE              FT  Principal Data Scientist   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
5          2023               SE              FT         Applied Scientist   
6          2023               SE              FT         Applied Scientist   
...         ...              ...             ...                       ...   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

         salary salary_currency  salary_in_usd employee_residence  \
0       80000.0             EUR          85847                 ES   
3      175000.0             USD         175000                 CA   
4      120000.0             USD         120000                 CA   
5      222200.0             USD         222200                 US   
6      136000.0             USD         136000                 US   
...         ...             ...            ...                ...   
3750   412000.0             USD         412000                 US   
3751   151000.0             USD         151000                 US   
3752   105000.0             USD         105000                 US   
3753   100000.0             USD         100000                 US   
3754  7000000.0             INR          94665                 IN   

      remote_ratio company_location company_size  
0              100               ES            L  
3              100               CA            M  
4              100               CA            M  
5                0               US            L  
6                0               US            L  
...            ...              ...          ...  
3750           100               US            L  
3751           100               US            L  
3752           100               US            S  
3753           100               US            L  
3754            50               IN            L  

[3142 rows x 11 columns]

In [19]:
#Фильтр колонки 3
df[df["remote_ratio"] < 30]

work_year experience_level employment_type                 job_title  \
5          2023               SE              FT         Applied Scientist   
6          2023               SE              FT         Applied Scientist   
7          2023               SE              FT            Data Scientist   
8          2023               SE              FT            Data Scientist   
9          2023               SE              FT            Data Scientist   
...         ...              ...             ...                       ...   
3730       2020               SE              FT            Data Scientist   
3736       2021               MI              FT              Data Analyst   
3737       2021               MI              FT              Data Analyst   
3738       2021               MI              FT            Data Scientist   
3745       2021               SE              FT  Director of Data Science   

        salary salary_currency  salary_in_usd employee_residence  \
5     222200.0             USD         222200                 US   
6     136000.0             USD         136000                 US   
7     219000.0             USD         219000                 CA   
8     141000.0             USD         141000                 CA   
9     147100.0             USD         147100                 US   
...        ...             ...            ...                ...   
3730   80000.0             EUR          91237                 AT   
3736   75000.0             USD          75000                 US   
3737   62000.0             USD          62000                 US   
3738   73000.0             USD          73000                 US   
3745  168000.0             USD         168000                 JP   

      remote_ratio company_location company_size  
5                0               US            L  
6                0               US            L  
7                0               CA            M  
8                0               CA            M  
9                0               US            M  
...            ...              ...          ...  
3730             0               AT            S  
3736             0               US            L  
3737             0               US            L  
3738             0               US            L  
3745             0               JP            S  

[1923 rows x 11 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [25]:
#df.groupby(["job_title"]).mean()
df.groupby(['job_title']).mean(numeric_only=True)
#df["job_title"].mean()

work_year         salary  salary_in_usd  \
job_title                                                                  
3D Computer Vision Researcher       2021.5       120000.0   21352.250000   
AI Developer                   2022.727273  137181.818182  136666.090909   
AI Programmer                       2022.5        55000.0   55000.000000   
AI Scientist                     2021.6875       275312.5  110120.875000   
Analytics Engineer             2022.446602  151352.621359  152368.631068   
...                                    ...            ...            ...   
Research Engineer              2022.864865  162752.810811  163108.378378   
Research Scientist                  2022.5  160768.890244  161214.195122   
Software Data Engineer              2023.0        75000.0   62510.000000   
Staff Data Analyst                  2020.0        15000.0   15000.000000   
Staff Data Scientist                2021.0       105000.0  105000.000000   

                               remote_ratio  
job_title                                    
3D Computer Vision Researcher     50.000000  
AI Developer                      31.818182  
AI Programmer                     50.000000  
AI Scientist                      71.875000  
Analytics Engineer                54.368932  
...                                     ...  
Research Engineer                 18.918919  
Research Scientist                27.439024  
Software Data Engineer            75.000000  
Staff Data Analyst                 0.000000  
Staff Data Scientist             100.000000  

[93 rows x 4 columns]

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.